File: backend/app/schemas/organization.py
"""组织架构相关 Schemas"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime, date
from decimal import Decimal


# ===== 催收机构 Schemas =====
class CollectionAgencyBase(BaseModel):
    """催收机构基础Schema"""
    tenant_id: int = Field(..., description="所属甲方ID")
    agency_code: str = Field(..., max_length=100, description="机构编码")
    agency_name: str = Field(..., max_length=200, description="机构名称")
    agency_name_en: Optional[str] = Field(None, max_length=200, description="机构名称（英文）")
    contact_person: Optional[str] = Field(None, max_length=100, description="联系人")
    contact_phone: Optional[str] = Field(None, max_length=50, description="联系电话")
    contact_email: Optional[str] = Field(None, max_length=100, description="联系邮箱")
    address: Optional[str] = Field(None, description="机构地址")
    description: Optional[str] = Field(None, description="机构描述")
    timezone: Optional[int] = Field(None, ge=-12, le=14, description="时区偏移量（UTC偏移，范围-12到+14）")
    agency_type: str = Field('real', description="机构类型：real=真实机构，virtual=虚拟机构")
    sort_order: int = Field(0, description="排序顺序")
    is_active: bool = Field(True, description="是否启用")


class CollectionAgencyCreate(CollectionAgencyBase):
    """创建催收机构"""
    pass


class CollectionAgencyUpdate(BaseModel):
    """更新催收机构"""
    agency_name: Optional[str] = Field(None, max_length=200)
    agency_name_en: Optional[str] = Field(None, max_length=200)
    contact_person: Optional[str] = Field(None, max_length=100)
    contact_phone: Optional[str] = Field(None, max_length=50)
    contact_email: Optional[str] = Field(None, max_length=100)
    address: Optional[str] = None
    description: Optional[str] = None
    timezone: Optional[int] = Field(None, ge=-12, le=14)
    agency_type: Optional[str] = Field(None, description="机构类型：real=真实机构，virtual=虚拟机构")
    sort_order: Optional[int] = None
    is_active: Optional[bool] = None


class CollectionAgency(CollectionAgencyBase):
    """催收机构响应"""
    id: int
    team_count: int = Field(0, description="小组数量")
    collector_count: int = Field(0, description="催员数量")
    case_count: int = Field(0, description="案件数量")
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True


# ===== 小组群 Schemas =====
class TeamGroupBase(BaseModel):
    """小组群基础Schema"""
    tenant_id: int = Field(..., description="所属甲方ID")
    agency_id: int = Field(..., description="所属催收机构ID")
    group_code: str = Field(..., max_length=100, description="小组群编码")
    group_name: str = Field(..., max_length=200, description="小组群名称")
    group_name_en: Optional[str] = Field(None, max_length=200, description="小组群名称（英文）")
    description: Optional[str] = Field(None, description="小组群描述")
    sort_order: int = Field(0, description="排序顺序")
    is_active: bool = Field(True, description="是否启用")


class TeamGroupCreate(BaseModel):
    """创建小组群"""
    tenant_id: int = Field(..., description="所属甲方ID")
    agency_id: int = Field(..., description="所属催收机构ID")
    group_code: str = Field(..., max_length=100, description="小组群编码")
    group_name: str = Field(..., max_length=200, description="小组群名称")
    group_name_en: Optional[str] = Field(None, max_length=200, description="小组群名称（英文）")
    description: Optional[str] = Field(None, description="小组群描述")
    sort_order: int = Field(0, description="排序顺序")
    is_active: bool = Field(True, description="是否启用")
    # SPV管理员账号信息
    spv_account_name: str = Field(..., max_length=100, description="SPV账号名称")
    spv_login_id: str = Field(..., max_length=100, description="SPV登录ID")
    spv_email: str = Field(..., max_length=100, description="SPV邮箱")
    spv_password: str = Field(..., max_length=50, description="SPV初始密码")
    spv_mobile: Optional[str] = Field(None, max_length=50, description="SPV手机号码")
    spv_remark: Optional[str] = Field(None, description="SPV备注")


class TeamGroupUpdate(BaseModel):
    """更新小组群"""
    group_name: Optional[str] = Field(None, max_length=200)
    group_name_en: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = None
    sort_order: Optional[int] = None
    is_active: Optional[bool] = None


class TeamGroup(TeamGroupBase):
    """小组群响应"""
    id: int
    agency_name: Optional[str] = Field(None, description="机构名称")
    spv_account_name: Optional[str] = Field(None, description="小组群长SPV姓名")
    spv_login_id: Optional[str] = Field(None, description="SPV登录ID")
    team_count: int = Field(0, description="小组数量")
    collector_count: int = Field(0, description="催员数量")
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True


# ===== 催收小组 Schemas =====
class CollectionTeamBase(BaseModel):
    """催收小组基础Schema"""
    agency_id: int = Field(..., description="所属催收机构ID")
    team_group_id: int = Field(..., description="所属小组群ID（必选）")
    queue_id: int = Field(..., description="关联的催收队列ID（必选）")
    team_code: str = Field(..., max_length=100, description="小组编码")
    team_name: str = Field(..., max_length=200, description="小组名称")
    team_name_en: Optional[str] = Field(None, max_length=200, description="小组名称（英文）")
    team_leader_id: Optional[int] = Field(None, description="组长ID")
    team_type: Optional[str] = Field(None, max_length=50, description="小组类型")
    description: Optional[str] = Field(None, description="小组描述")
    max_case_count: int = Field(0, description="最大案件数量")
    sort_order: int = Field(0, description="排序顺序")
    is_active: bool = Field(True, description="是否启用")


class CollectionTeamCreate(CollectionTeamBase):
    """创建催收小组"""
    pass


class CollectionTeamUpdate(BaseModel):
    """更新催收小组"""
    team_group_id: Optional[int] = Field(None, description="所属小组群ID")
    queue_id: Optional[int] = Field(None, description="关联的催收队列ID")
    team_name: Optional[str] = Field(None, max_length=200)
    team_name_en: Optional[str] = Field(None, max_length=200)
    team_leader_id: Optional[int] = None
    team_type: Optional[str] = Field(None, max_length=50)
    description: Optional[str] = None
    max_case_count: Optional[int] = None
    sort_order: Optional[int] = None
    is_active: Optional[bool] = None


class CollectionTeam(CollectionTeamBase):
    """催收小组响应"""
    id: int
    tenant_id: int
    tenant_name: Optional[str] = Field(None, description="甲方名称")
    agency_name: Optional[str] = Field(None, description="机构名称")
    team_group_name: Optional[str] = Field(None, description="小组群名称")
    queue_name: Optional[str] = Field(None, description="催收队列名称")
    team_leader_name: Optional[str] = Field(None, description="组长姓名")
    collector_count: int = Field(0, description="催员数量")
    case_count: int = Field(0, description="案件数量")
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True


# ===== 催员 Schemas =====
class CollectorBase(BaseModel):
    """催员基础Schema"""
    team_id: int = Field(..., description="所属小组ID")
    user_id: int = Field(..., description="关联用户ID")
    collector_code: str = Field(..., max_length=100, description="催员编码")
    collector_name: str = Field(..., max_length=100, description="催员姓名")
    mobile_number: Optional[str] = Field(None, max_length=50, description="手机号码")
    email: Optional[str] = Field(None, max_length=100, description="邮箱")
    employee_no: Optional[str] = Field(None, max_length=50, description="工号")
    collector_level: Optional[str] = Field(None, max_length=50, description="催员等级")
    max_case_count: int = Field(100, description="最大案件数量")
    specialties: Optional[List[str]] = Field(None, description="擅长领域")
    hire_date: Optional[date] = Field(None, description="入职日期")
    is_active: bool = Field(True, description="是否启用")


class CollectorCreate(CollectorBase):
    """创建催员"""
    pass


class CollectorUpdate(BaseModel):
    """更新催员"""
    team_id: Optional[int] = None
    collector_name: Optional[str] = Field(None, max_length=100)
    mobile_number: Optional[str] = Field(None, max_length=50)
    email: Optional[str] = Field(None, max_length=100)
    employee_no: Optional[str] = Field(None, max_length=50)
    collector_level: Optional[str] = Field(None, max_length=50)
    max_case_count: Optional[int] = None
    specialties: Optional[List[str]] = None
    status: Optional[str] = Field(None, max_length=50)
    is_active: Optional[bool] = None


class CollectorStatistics(BaseModel):
    """催员统计数据"""
    today_assigned: int = Field(0, description="今日分配案件数")
    week_assigned: int = Field(0, description="本周分配案件数")
    month_settled: int = Field(0, description="本月结清案件数")
    month_amount_collected: Decimal = Field(0, description="本月回款金额")
    month_contact_count: int = Field(0, description="本月接触次数")
    case_status_distribution: dict = Field({}, description="案件状态分布")


class Collector(CollectorBase):
    """催员响应"""
    id: int
    agency_name: Optional[str] = Field(None, description="机构名称")
    team_name: Optional[str] = Field(None, description="小组名称")
    current_case_count: int = Field(0, description="当前案件数量")
    performance_score: Optional[Decimal] = Field(None, description="绩效评分")
    status: str = Field("active", description="状态")
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True


class CollectorDetail(Collector):
    """催员详情（包含统计）"""
    statistics: Optional[CollectorStatistics] = None


# ===== 案件分配 Schemas =====
class CaseAssignmentBase(BaseModel):
    """案件分配基础Schema"""
    case_ids: List[int] = Field(..., description="案件ID列表")
    reason: Optional[str] = Field(None, max_length=200, description="分配原因")
    remarks: Optional[str] = Field(None, description="备注")


class AssignToAgency(CaseAssignmentBase):
    """分配案件到机构"""
    agency_id: int = Field(..., description="目标机构ID")


class AssignToTeam(CaseAssignmentBase):
    """分配案件到小组"""
    team_id: int = Field(..., description="目标小组ID")


class AssignToCollector(CaseAssignmentBase):
    """分配案件到催员"""
    collector_id: int = Field(..., description="目标催员ID")


class AutoAssignRequest(CaseAssignmentBase):
    """自动分配请求"""
    assignment_rule: str = Field("balanced", description="分配规则：balanced/random/skill_match")
    target_level: str = Field("collector", description="分配层级：agency/team/collector")
    agency_id: Optional[int] = Field(None, description="指定机构（可选）")
    team_id: Optional[int] = Field(None, description="指定小组（可选）")


class TransferCaseRequest(BaseModel):
    """转移案件请求"""
    case_ids: List[int] = Field(..., description="案件ID列表")
    from_collector_id: int = Field(..., description="原催员ID")
    to_collector_id: int = Field(..., description="目标催员ID")
    reason: str = Field(..., max_length=200, description="转移原因")
    remarks: Optional[str] = Field(None, description="备注")


class ReclaimCaseRequest(BaseModel):
    """回收案件请求"""
    case_ids: List[int] = Field(..., description="案件ID列表")
    reason: str = Field(..., max_length=200, description="回收原因")
    target_level: str = Field("team", description="回收到的层级：team/agency")


class AssignmentResult(BaseModel):
    """分配结果"""
    case_id: int
    assigned_to_collector_id: Optional[int] = None
    assigned_to_collector_name: Optional[str] = None
    assigned_to_team_id: Optional[int] = None
    assigned_to_team_name: Optional[str] = None
    assigned_to_agency_id: Optional[int] = None
    assigned_to_agency_name: Optional[str] = None


class AssignmentResponse(BaseModel):
    """分配响应"""
    assigned_count: int = Field(..., description="成功分配数量")
    assignments: List[AssignmentResult] = Field(..., description="分配详情")


# ===== 机构作息时间 Schemas =====
class TimeSlot(BaseModel):
    """时间段"""
    start: str = Field(..., description="开始时间（格式：HH:MM，24小时制）")
    end: str = Field(..., description="结束时间（格式：HH:MM，24小时制）")


class AgencyWorkingHoursBase(BaseModel):
    """机构作息时间基础Schema"""
    agency_id: int = Field(..., description="机构ID")
    day_of_week: int = Field(..., ge=0, le=6, description="星期几（0=周一，6=周日）")
    time_slots: List[TimeSlot] = Field(..., max_length=5, description="时间段列表（最多5个）")


class AgencyWorkingHoursCreate(AgencyWorkingHoursBase):
    """创建机构作息时间"""
    pass


class AgencyWorkingHoursUpdate(BaseModel):
    """更新机构作息时间"""
    time_slots: List[TimeSlot] = Field(..., max_length=5, description="时间段列表（最多5个）")


class AgencyWorkingHoursResponse(AgencyWorkingHoursBase):
    """机构作息时间响应"""
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AgencyWorkingHoursBatchUpdate(BaseModel):
    """批量更新机构作息时间"""
    working_hours: List[AgencyWorkingHoursCreate] = Field(..., description="作息时间列表（7天）")


# ===== 案件分配历史 Schemas =====
class CaseAssignmentHistory(BaseModel):
    """案件分配历史"""
    id: int
    case_id: int
    assignment_type: str = Field(..., description="分配类型")
    from_agency_name: Optional[str] = None
    to_agency_name: Optional[str] = None
    from_team_name: Optional[str] = None
    to_team_name: Optional[str] = None
    from_collector_name: Optional[str] = None
    to_collector_name: Optional[str] = None
    reason: Optional[str] = None
    assigned_by_name: Optional[str] = None
    assigned_at: datetime
    remarks: Optional[str] = None

    class Config:
        orm_mode = True

File: frontend/src/types/organization.ts
/**
 * 组织架构相关类型定义
 */

// ===== 催收机构 =====
export interface CollectionAgency {
  id: number
  tenant_id: number
  agency_code: string
  agency_name: string
  agency_name_en?: string
  contact_person?: string
  contact_phone?: string
  contact_email?: string
  address?: string
  description?: string
  timezone?: number
  agency_type: string  // 'real' | 'virtual'
  sort_order: number
  is_active: boolean
  team_count: number
  collector_count: number
  case_count: number
  created_at: string
  updated_at: string
}

export interface CollectionAgencyCreate {
  tenant_id: number
  agency_code: string
  agency_name: string
  agency_name_en?: string
  contact_person?: string
  contact_phone?: string
  contact_email?: string
  address?: string
  description?: string
  agency_type?: string  // 'real' | 'virtual'
  sort_order?: number
  is_active?: boolean
}

export interface CollectionAgencyUpdate {
  agency_name?: string
  agency_name_en?: string
  contact_person?: string
  contact_phone?: string
  contact_email?: string
  address?: string
  description?: string
  timezone?: number
  agency_type?: string  // 'real' | 'virtual'
  sort_order?: number
  is_active?: boolean
}

// ===== 小组群 =====
export interface TeamGroup {
  id: number
  tenant_id: number
  agency_id: number
  group_code: string
  group_name: string
  group_name_en?: string
  description?: string
  sort_order: number
  is_active: boolean
  agency_name?: string
  spv_account_name?: string
  spv_login_id?: string
  team_count: number
  collector_count: number
  created_at: string
  updated_at: string
}

export interface TeamGroupCreate {
  tenant_id: number
  agency_id: number
  group_code: string
  group_name: string
  group_name_en?: string
  description?: string
  sort_order?: number
  is_active?: boolean
  // SPV管理员账号信息
  spv_account_name: string
  spv_login_id: string
  spv_email: string
  spv_password: string
  spv_remark?: string
}

export interface TeamGroupUpdate {
  group_name?: string
  group_name_en?: string
  description?: string
  sort_order?: number
  is_active?: boolean
}

// ===== 催收小组 =====
export interface CollectionTeam {
  id: number
  agency_id: number
  team_group_id?: number
  queue_id: number
  team_code: string
  team_name: string
  team_name_en?: string
  team_leader_id?: number
  team_type?: string
  description?: string
  max_case_count: number
  sort_order: number
  is_active: boolean
  agency_name?: string
  team_group_name?: string
  queue_name?: string
  team_leader_name?: string
  collector_count: number
  case_count: number
  created_at: string
  updated_at: string
}

export interface CollectionTeamCreate {
  agency_id: number
  team_group_id?: number
  queue_id: number
  team_code: string
  team_name: string
  team_name_en?: string
  team_leader_id?: number
  team_type?: string
  description?: string
  max_case_count?: number
  sort_order?: number
  is_active?: boolean
}

export interface CollectionTeamUpdate {
  team_group_id?: number
  queue_id?: number
  team_name?: string
  team_name_en?: string
  team_leader_id?: number
  team_type?: string
  description?: string
  max_case_count?: number
  sort_order?: number
  is_active?: boolean
}

// ===== 催员 =====
export interface Collector {
  id: number
  team_id: number
  user_id: number
  collector_code: string
  collector_name: string
  mobile_number?: string
  email?: string
  employee_no?: string
  collector_level?: string
  max_case_count: number
  current_case_count: number
  specialties?: string[]
  performance_score?: number
  status: string
  hire_date?: string
  is_active: boolean
  agency_name?: string
  team_name?: string
  created_at: string
  updated_at: string
}

export interface CollectorCreate {
  team_id: number
  user_id: number
  collector_code: string
  collector_name: string
  mobile_number?: string
  email?: string
  employee_no?: string
  collector_level?: string
  max_case_count?: number
  specialties?: string[]
  hire_date?: string
  is_active?: boolean
}

export interface CollectorUpdate {
  team_id?: number
  collector_name?: string
  mobile_number?: string
  email?: string
  employee_no?: string
  collector_level?: string
  max_case_count?: number
  specialties?: string[]
  status?: string
  is_active?: boolean
}

export interface CollectorStatistics {
  today_assigned: number
  week_assigned: number
  month_settled: number
  month_amount_collected: number
  month_contact_count: number
  case_status_distribution: Record<string, number>
}

export interface CollectorDetail extends Collector {
  statistics?: CollectorStatistics
}

// ===== 案件分配 =====
export interface AssignToAgency {
  case_ids: number[]
  agency_id: number
  reason?: string
  remarks?: string
}

export interface AssignToTeam {
  case_ids: number[]
  team_id: number
  reason?: string
  remarks?: string
}

export interface AssignToCollector {
  case_ids: number[]
  collector_id: number
  reason?: string
  remarks?: string
}

export interface AutoAssignRequest {
  case_ids: number[]
  assignment_rule: 'balanced' | 'random' | 'skill_match'
  target_level: 'agency' | 'team' | 'collector'
  agency_id?: number
  team_id?: number
  reason?: string
  remarks?: string
}

export interface TransferCaseRequest {
  case_ids: number[]
  from_collector_id: number
  to_collector_id: number
  reason: string
  remarks?: string
}

export interface ReclaimCaseRequest {
  case_ids: number[]
  reason: string
  target_level: 'team' | 'agency'
}

export interface AssignmentResult {
  case_id: number
  assigned_to_collector_id?: number
  assigned_to_collector_name?: string
  assigned_to_team_id?: number
  assigned_to_team_name?: string
  assigned_to_agency_id?: number
  assigned_to_agency_name?: string
}

export interface AssignmentResponse {
  assigned_count: number
  assignments: AssignmentResult[]
}

// ===== 案件分配历史 =====
export interface CaseAssignmentHistory {
  id: number
  case_id: number
  assignment_type: string
  from_agency_name?: string
  to_agency_name?: string
  from_team_name?: string
  to_team_name?: string
  from_collector_name?: string
  to_collector_name?: string
  reason?: string
  assigned_by_name?: string
  assigned_at: string
  remarks?: string
}

// ===== 机构作息时间 =====
export interface TimeSlot {
  start: string  // 格式：HH:MM（24小时制）
  end: string    // 格式：HH:MM（24小时制）
}

export interface AgencyWorkingHours {
  id: number
  agency_id: number
  day_of_week: number  // 0=周一，6=周日
  time_slots: TimeSlot[]
  created_at: string
  updated_at: string
}

export interface AgencyWorkingHoursUpdate {
  time_slots: TimeSlot[]
}

export interface AgencyWorkingHoursBatchUpdate {
  working_hours: Array<{
    agency_id: number
    day_of_week: number
    time_slots: TimeSlot[]
  }>
}

